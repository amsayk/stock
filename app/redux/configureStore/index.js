// @flow

import { createStore, applyMiddleware, compose } from 'redux';

import { APP_NAME, DEBUG, INIT } from 'vars';

import pick from 'lodash.pick';

import array from 'redux/middleware/array';
import bufferActions from 'redux/middleware/buffer-actions';

import { createHashHistory } from 'history';
import { routerMiddleware, routerActions } from 'react-router-redux';
import locationHelperBuilder from 'redux-auth-wrapper/history4/locationHelper';

import { client as apolloClient } from 'apollo';

import { createSnackbarController } from 'components/Snackbar';

import thunk from 'redux-thunk';

import rootReducer from 'redux/reducers';

const history = createHashHistory();

const configureStore = (initialState?: any) => {
  // ======================================================
  // Middleware Configuration
  // ======================================================
  const middlewares = [
    bufferActions(),
    array,
    thunk.withExtraArgument({
      client: apolloClient,
      history,
      get snackbar() {
        return (
          store.snackbar || (store.snackbar = createSnackbarController(store))
        );
      },
    }),
    routerMiddleware(history),
  ];

  // Redux DevTools Configuration
  const actionCreators = {
    ...routerActions,
  };

  const composeEnhancers =
    (__DEV__ || DEBUG) && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
      ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
          name: APP_NAME,
          routerActions,
        })
      : compose;

  // ======================================================
  // Store Enhancers
  // ======================================================
  const enhancers = [applyMiddleware(...middlewares)];

  const enhancer = composeEnhancers(...enhancers);

  const store = createStore(rootReducer, enhancer);

  // Snackbar
  store.snackbar = createSnackbarController(store);

  // Initialize
  store.dispatch({ type: INIT });

  if (__DEV__) {
    if (module.hot) {
      module.hot.accept('../reducers', () => {
        store.replaceReducer(require('../reducers'));
      });
    }
  }

  return store;
};

export default {
  configureStore,
  history,
  locationHelper: locationHelperBuilder({}),
};
