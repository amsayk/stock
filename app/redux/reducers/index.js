import user from './user/reducer';
import app from './app/reducer';
import snackbar from './snackbar/reducer';
import notification from './notification/reducer';
import scrolling from './scrolling/reducer';
import toastr from './toastr/reducer';
import intl from './intl/reducer';

import { USER_LOGGED_OUT } from './user/constants';

import { combineReducers } from 'redux-immutable';

import { INIT } from 'vars';

import { routerReducer as router } from 'react-router-redux';

import { reducer as formReducer, actionTypes } from 'redux-form/immutable';

const reducers = {
  intl,
  app,
  notification,
  scrolling,
  user,
  snackbar,
  toastr,
  form: formReducer.plugin({
    login: (state, action) => {
      switch (action.type) {
        case USER_LOGGED_OUT:
          return state
            ? state.mergeDeep({
                values: {
                  password: undefined,
                },
              })
            : state;
        case actionTypes.SUBMIT_START: // Clear errors
          return state
            ? state.merge({
                error: null,
              })
            : state;
        default:
          return state;
      }
    },
    changePassword: (state, action) => {
      switch (action.type) {
        case actionTypes.SET_SUBMIT_SUCCEEDED: // Clear passwords on success
          return state
            ? state.merge({
                values: {},
              })
            : state;
        default:
          return state;
      }
    },
    changeEmail: (state, action) => {
      switch (action.type) {
        case actionTypes.SET_SUBMIT_SUCCEEDED: // Clear email on success
          return state
            ? state.merge({
                values: {},
              })
            : state;
        default:
          return state;
      }
    },
  }),
};

export default combineReducers({
  router,
  ...reducers,
});
