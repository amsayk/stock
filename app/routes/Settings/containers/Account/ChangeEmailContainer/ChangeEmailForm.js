import React from 'react';
import T from 'prop-types';
import { compose } from 'redux';

import { withApollo } from 'react-apollo';

import refreshCurrentUser from 'utils/refreshCurrentUser';

import {
  reduxForm,
  Field,
  propTypes as reduxFormPropTypes,
  SubmissionError,
} from 'redux-form/immutable';

import isEmpty from 'isEmpty';

import { intlShape, FormattedMessage } from 'react-intl';

import messages from 'routes/Settings/messages';

import style from 'routes/Settings/styles';

import NewEmailField from 'routes/Settings/components/NewEmailField';

import MUTATION from './changeEmail.mutation.graphql';

export class ChangeEmailForm extends React.Component {
  constructor(props, context) {
    super(props, context);

    this.onKeyDown = this._onKeyDown.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }
  _onKeyDown(e) {
    if (e.key === 'Enter' && e.shiftKey === false) {
      const submit = this.props.handleSubmit(this.onSubmit);
      submit();
    }
  }

  async onSubmit(data) {
    const {
      data: { changeEmail: { user, errors } },
    } = await this.props.client.mutate({
      mutation: MUTATION,
      variables: {
        payload: {
          email: data.get('email'),
        },
      },
    });

    if (!isEmpty(errors)) {
      throw new SubmissionError(errors);
    }

    // currentUser has changed, refresh.
    await refreshCurrentUser(user, this.props.dispatch);

    const { intl } = this.props;
    const { snackbar } = this.context;
    if (snackbar) {
      snackbar.show({
        message: 'Succ√®s', // intl.formatMessage(messages.emailChangeSuccessNotification),
        duration: 2000,
      });
    }
  }

  render() {
    const {
      intl,
      user,
      handleSubmit,
      pristine,
      submitting,
      invalid,
    } = this.props;
    return (
      <div className={style.content}>
        <h1 className={style.changeEmailFormHeading}>
          {intl.formatMessage(messages.titleChangeEmail)}
        </h1>
        <p className={style.currentEmailIntro}>
          <FormattedMessage
            {...messages.currentEmailIntro}
            values={{
              email: <strong>{user.email}</strong>,
            }}
          />
        </p>
        <div className={style.form}>
          <Field
            name="email"
            component={NewEmailField}
            label={intl.formatMessage(messages.labelNewEmail)}
            onKeyDown={this.onKeyDown}
          />
          <div className={style.formGroup}>
            <div className={style.offset2} />
            <button
              onClick={handleSubmit(this.onSubmit)}
              disabled={submitting || invalid}
              className={style.saveButton}
            >
              {intl.formatMessage(messages.changeEmail)}
            </button>
          </div>
        </div>
      </div>
    );
  }
}

ChangeEmailForm.contextTypes = {
  snackbar: T.shape({
    show: T.func.isRequired,
  }),
};

ChangeEmailForm.defaultProps = {};

ChangeEmailForm.propTypes = {
  ...reduxFormPropTypes,
  intl: intlShape.isRequired,
  user: T.shape({
    email: T.string,
  }).isRequired,
};

const Form = reduxForm({
  form: 'changeEmail',
  enableReinitialize: true,
  keepDirtyOnReinitialize: false,
});

export default compose(withApollo, Form)(ChangeEmailForm);
