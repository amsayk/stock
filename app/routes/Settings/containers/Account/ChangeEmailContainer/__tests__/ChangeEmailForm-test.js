import React from 'react';
import { compose } from 'redux';
import configureMockStore from 'redux-mock-store';
import { reduxForm } from 'redux-form/immutable';
import { IntlProvider, injectIntl } from 'react-intl';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

import { fromJS } from 'immutable';

import { ChangeEmailForm } from '../ChangeEmailForm';

const client = {};

const user = {
  email: 'email',
};

const Decorated = compose(
  injectIntl,
  reduxForm({
    form: 'testForm',
  }),
)(ChangeEmailForm);

const middlewares = [];
const mockStore = configureMockStore(middlewares);

describe('change email form', () => {
  const store = mockStore(
    fromJS({
      form: {
        testForm: {},
      },
    }),
  );

  it('should render the snapshot', () => {
    const tree = renderer
      .create(
        <IntlProvider
          defaultLocale={'en'}
          locale={'en'}
          messages={{}}
          formats={{}}
        >
          <Provider store={store}>
            <Decorated user={user} client={client} />
          </Provider>
        </IntlProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
