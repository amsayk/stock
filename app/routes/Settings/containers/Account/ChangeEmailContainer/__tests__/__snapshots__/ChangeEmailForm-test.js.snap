exports[`change email form should render the snapshot 1`] = `
<div
  className="content">
  <h1
    className="changeEmailFormHeading">
    Change email address
  </h1>
  <p
    className="currentEmailIntro">
    <span>
      Your email address is currently 
      <strong>
        email
      </strong>
      .
    </span>
  </p>
  <div
    className="form">
    <div
      className="formGroup">
      <label
        className="label"
        htmlFor="email">
        New email address
      </label>
      <div
        className="inputWrapper">
        <input
          autoComplete="off"
          autoFocus={true}
          className="control"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          type="text"
          value="" />
        <div />
      </div>
    </div>
    <button
      className="saveButton"
      disabled={true}
      onClick={[Function]}>
      Change email
    </button>
  </div>
</div>
`;
